### PR TITLE
Allowing consumers to use their own queues

### PR DESCRIPTION
**Pull request by bsrz**
We recently experienced a bug where our UI wasn't able to update asynchronously because of scrolling, upside down table views, etc... I'll spare you the details.
The point is that is was absolutely impossible to make parsing call synchronous because the completion closure was going to be called on the main thread no matter what.
This is the same queue we wanted to block because the call originated from a UI component.

TL; DR;
This PR exposes which queue to run the work and which queue to call the completion closure.
This makes it possible to make the call synchronous if needed.

### CODE REVIEW

**Comment by bsrz** [comment]
@daltoniam bump

**Comment by daltoniam** [comment]
looks good and thanks for the detailed write up!
